src/os_posix/os_fs.c

```diff
@@ -501,9 +563,31 @@ __posix_file_truncate(WT_FILE_HANDLE *file_handle, WT_SESSION *wt_session, wt_of
     session = (WT_SESSION_IMPL *)wt_session;
     pfh = (WT_FILE_HANDLE_POSIX *)file_handle;
 
+    if (S2C(session)->mmap) {
+        __wt_verbose(session, WT_VERB_FILEOPS, "%s, file-truncate: size=%" PRId64
+                                               ","
+                                               "mapped size=%" PRIu64 "\n",
+          file_handle->name, len, (uint64_t)pfh->mmap_size);
+
+        if (pfh->mmap_file_mappable && (wt_off_t)pfh->mmap_size != len)
+            __wt_drain_mmap_users(file_handle, wt_session);
+    }
+
     WT_SYSCALL_RETRY(ftruncate(pfh->fd, len), ret);
-    if (ret == 0)
+    if (ret == 0) {
+        if (S2C(session)->mmap) {
+            /* Remap the region with the new size */
+            if (pfh->mmap_file_mappable && (wt_off_t)pfh->mmap_size != len) {
+                ret = __wt_remap_region(file_handle, wt_session);
+                if (ret) {
+                    S2C(session)->mmap = false;
+                    __wt_err(session, ret, "Could not remap file. Mmap option disabled.");
+                }
+                WT_STAT_CONN_INCRV(session, block_remap_region_trunc, 1);
+            }
+        }
         return (0);
+    }
```

Comment:
Noting delay of error handling, probably not necessary and leads to additional indentation.